### ISSUE TITLE
Outlook for Tendermint Databases [Tracking issue]

### ISSUE THREAD

**Issue by cmwaters** [tendermint/tendermint#6032]
## Summary

There have been a few discussions already surrounding the topic. This is an issue that tries to capture the current landscape and offer some guidance towards heading forwards. Essentially the crux of the topic is more of a question: How involved does Tendermint want to be with the database layer that it operates above? The core value of Tendermint resides in BFT state machine replication. This means modules such as consensus, p2p, mempool and so forth. Because of this focus, we should slowly reduce the large maintenance surface area that Tendermint currently holds with the database layer. In its place, we should look to settle for leading and established technologies that we can leverage off. 

Specifically that points three issues:

- Reduce the amount of databases supported, eventually converging to a singe database that best fits Tendermint's use case.
- Implement an established query engine to use for the Tx Indexer (such as sqlite or postgres)
- Reassess the write ahead log

## Converge to a single database

Currently, tm-db supports five different databases (GoLevelDB, LevelDB, BoltDB, RocksDB and BadgerDB). This offers a greater degree of flexibility to application developers but it's not clear whether such flexibility provides the necessary gain to offset the maintenance burden of providing continual support. Having this wrapped around a common interface restricts that amount of value we can extract from each of the respective databases and means that we can't exploit some of the invariants of Tendermint (append only, high write volume and so forth). 

In order to move forward, we will need to undergo a process of comparing the various technologies out there, writing them all up in a document, conversing with SDK and other users and eventually reach a consensus on the db engine of choice. 

Following from this, another document, this time more implementation focused and most likely an ADR, will need to be sketched out. Given a single database engine we will need to review how it should now interface with Tendermint. This could likely lead to a large rewrite of some of the internal code. Another question that would arise then would be: Do we want to use only one or perhaps two databases instead of four different ones. This may offer greater atomicity of transactions.

## Implement an established query engine for the Tx Indexer

The Tx Indexer is a secondary index to query and find particular Txs. Whilst the current implementation of a kv store suffices, there is already a wealth of highly performant query engines that we should consider using (sqlite, postgres). This will allow us to piggy back off the latest technology whilst only needing to maintain a much smaller codebase (the wrapper around the engine). 

## Reassess the write ahead log

The WAL, records all messages that the consensus engine produces so that in the event of a crash, when the node starts up again it can replay all the messages back to the point of failure.  This is another, although perhaps smaller, area where we might be able to reduce the amount of code we need to upkeep. It seems, at least to me, that we only need a write ahead log for the votes that we sign in order to prevent double signing yet for everything else we could rely on the existing "catch up" mechanism to receive all the votes and blocks that have happened since the crash to restore the node to the most recent height. 

____

#### For Admin Use

- [ ] Not duplicate issue
- [ ] Appropriate labels applied
- [ ] Appropriate contributors tagged
- [ ] Contributor assigned/self-assigned

**Comment by marbar3778** [comment]
In conversation with the SDK team, they mentioned also dropping support for most of the DBs. It seems tm-db will be dropped from their codebase. This will help in our decision for tendermint and the tm-db repo.

**Comment by tessr** [comment]
> It seems tm-db will be dropped from their codebase. 

This means we could pull tm-db back into the main Tendermint repo, right?

**Comment by alexanderbez** [comment]
> > It seems tm-db will be dropped from their codebase.
> 
> This means we could pull tm-db back into the main Tendermint repo, right?

We should consider sun-setting it entirely IMHO.

**Comment by marbar3778** [comment]
> > > It seems tm-db will be dropped from their codebase.
> > 
> > 
> > This means we could pull tm-db back into the main Tendermint repo, right?
> 
> We should consider sun-setting it entirely IMHO.

Yes, I would be in favor of sun-setting tm-db as well.

**Comment by cmwaters** [comment]
Right but this means first making a decision on which db Tendermint want's to use (and therefore all the work required to reach that decision). Apart from cosmos sdk, are there any other major users of tm-db?

**Comment by alexanderbez** [comment]
Correct, the two can be done in parallel.

**Comment by ebuchman** [comment]
In favour of reducing maintenance burden re dbs. One note about the WAL, one of the original motivations for it was to have a means for deterministic replay of messages a node received, mostly for debugging and testing purposes.

**Comment by marbar3778** [comment]
I worked on a PR for lazyledger (https://github.com/lazyledger/lazyledger-core/pull/241) that reduces the supported dbs to badger.

We can upstream this work once we fix the timeout issue as part of this process.

**Comment by tessr** [comment]
Related issues:

* Concurrency problems in blockstore https://github.com/tendermint/tendermint/issues/5383 
* Older conversations about picking a db in tm-db https://github.com/tendermint/tm-db/issues/38 

Also, is it safe to say that the indexer issues will no longer be relevant once the streaming-only indexer (https://github.com/tendermint/tendermint/issues/6303) lands?

**Comment by marbar3778** [comment]
> Also, is it safe to say that the indexer issues will no longer be relevant once the streaming-only indexer (#6303) lands?

we will support the old indexer on top of the new ones for the time being. At a later date it will be irrelevant.

**Comment by faddat** [comment]
Is it safe / possible / advisable to completely remove the WAL?

It seems to limitthe use of Tendermint in embedded settings where we're storing data on SD Cards.

**Comment by alexanderbez** [comment]
I'm not sure it's entirely possible to remove the WAL @faddat, but what we _should_ do is rely on an appropriate DB for proper WAL functionality, instead of rolling our own.

The reason I think we can't remove it is because it's so tightly integrated into the safety and recovery mechanism of the consensus reactor.

**Comment by i-norden** [comment]
In ADR-040 we plan to support both BadgerDB and RocksDB, perhaps we could converge to supporting both of these instead of just Badger? The output of the work here will be the starting point to support ADR-040 with those two underlying DBs, so it could save some back-tracking.

**Comment by brennhill** [comment]
I looked into the postgres implementation of the indexer.  Right now for many functions the behavior of the indexer only works when using the KV store. Thus, if we move to something like Postgres, sections of code have to be re-written to still function with the new indexer.    I do think that requiring an external DB makes setting up tendermint more difficult, so both the KV and a DB should be supported. However, this should be done without disabling behavior (which is a surprising outcome).

For TM-DB, while all the databases listed are really nice, has there been any consideration of supporting DBs that are built to scale resources horizontally like CockroachDB or TiDB? The ability of a node to use a Database that has robust abilities to handle failover and scale seems like an easy win.

https://tableplus.com/blog/2018/10/cockroachdb-vs-postgresql-comparison.html

**Comment by creachadair** [comment]
> I looked into the postgres implementation of the indexer. Right now for many functions the behavior of the indexer only works when using the KV store. 

It's important to distinguish between the "indexer" and the "database" in the context of Tendermint. The database is integral to consensus, and is used (as noted) for the state and block stores, and the mempool. The indexer is optional, and does not share storage with the rest of the consensus engine.

For indexing, good issues to track include https://github.com/tendermint/tendermint/issues/7156 and https://github.com/tendermint/tendermint/issues/7135. 

For the database, I recommend starting with [RFC 001](https://github.com/tendermint/tendermint/blob/master/docs/rfc/rfc-001-storage-engine.rst).

Although it is possible, in principle, to merge these two concerns, that isn't the directly I think we are planning to go. The footprint and usage profiles of these concerns are different enough that it makes more sense to keep them separate.

**Comment by brennhill** [comment]
Ok, where can I start?  It sounds like some of these require a substantive re-architecture.

**Comment by faddat** [comment]
Hey guys, I think PebbleDB because:

* maintained
* used actively in / by cockroachb
* fast like rocks
* no cgo


If it wasn't pebble, then likely badger

**Comment by dwedul-figure** [comment]
Hey! I ran some of the sims test of our TM backed Provenance Blockchain using various database backends: goleveldb, cleveldb, rocksdb, and badgerdb. I wrote a about it too: https://medium.com/provenanceblockchain/the-database-backends-of-the-provenance-blockchain-c2a71b565859

tl;dr: `goleveldb` performed the best for all three sims that I ran.

Tests were run on a dedicated machine one at a time. I ran each test with each db 8 times.

Also, I never saw this happen during my tests, but in our [github actions](https://github.com/provenance-io/provenance/runs/6190019884?check_suite_focus=true) that run sims on Ubuntu using `rocksdb`, sometimes we get this error:
```
signal: illegal instruction (core dumped)
FAIL	github.com/provenance-io/provenance/app	0.508s
FAIL
make: *** [sims.mk:83: test-sim-simple] Error 1
Error: Process completed with exit code 2.
```
We currently have them using RocksDB version 6.29.5, but I know I've seen it in 6.29.4, 6.29.3, and at least a couple 6.28.x.

The external shared library is a major pain in the butt for rocksdb, and a minor pain for cleveldb.

**Comment by creachadair** [comment]
One meta-point to keep in mind for this issue is that we are mainly concerned here (and in RFC 001) with data storage for the consensus node, not for the application. 

I wouldn't be surprised at all if application workloads depend on different performance properties than the node itself does for state machine replication. For this issue the application is out of scope. Not that it doesn't matter, only that the two shouldn't be required to use the same storage engine.

**Comment by faddat** [comment]
@tychoish 

https://github.com/dgraph-io/badger/pull/1754

This does not make badger look so great, honestly.

That's the feature we need, and it just kind of.... sat.

I'm really interested in solving this well.  

@dwedul-figure sir, you and your team have done things that we've used at Notional several times, like the v44 wasm upgrade -- juno's original was based off yours.  Thank you. too bad I missed the 20 byte addresses :shame-face: still though, your code was super helpful as a guide at that time. 

I suspect that the reason goleveldb is "winning" is that if we compare, it's got the most optimized codebase.  I'm really interested in how you are doing those tests.

Given the above PR, I don't know if badger should really be considered, I would really like to end up using something whose team works along side us to tweak things.

I also have historically very-liked cockroach & their team which kind of leads me back to pebble. 

also for @dwedul-figure -- here's my most fancy tm-db branch:

https://github.com/notional-labs/tm-db/tree/hyped-badger

I hope to get pebble added to it soon, but every time I do pebble, I end up with inexplicable type errors.

**Comment by dwedul-figure** [comment]
@faddat Thank you for your kind words. I'd love to chat. We have both discord and slack channels for Provenance (see links at the bottom of our [Community](https://provenance.io/community) page). Feel free to email me too.

I kind of expect that the sims I used to get those times aren't indicative of real use, and possibly play to the strengths of `goleveldb`. I just used them because I needed the stats quickly, and the sims seemed like a somewhat decent benchmark tool. They at least do a bunch of writing (not sure how much reading they do though).

From my standpoint (as a blockchain developer), I agree with @faddat.
* Write time is mostly negligible as long as it is done faster than a new block is created. However, reducing write time will help nodes catch up faster (e.g. from a quicksync faster or after an upgrade), which is desirable.
* Query time is the primary concern. Query time for things used during TX processing are first priority. Query time for external queries is second priority. I figure the difference between the two is mostly caching related, but maybe not all.
* Requiring an external shared library is problematic. The different package management services used by various systems/people often provide different versions of these shared libraries, which can lead to problems and is often a barrier to entry.
* Being able to select different DB Backends is desirable. It allows the developers to more easily evaluate the different DB Backends in order to either define or recommend the DB Backend node operators should use.
* Being able to define all DB Backend types (including the Cosmos-SDK ones) with a single config entry, is also desirable for the sake of simplicity. But being able to define the backend for each DB is desirable for the sake of optimization. I just don't want the config to get too complex because I also have to help with customer (e.g. node operator) support, and complex configurations can be difficult in that respect.
* Total disk footprint is also a concern. A common node disk size is 500G. During an upgrade, a backup is often made, usually to the same disk. So if the total `data` footprint is more than 250G, there's a problem. With Provenance Blockchain, a mainnet node using BadgerDB takes up more than 250G of space. Most of that is in the application db (Cosmos-SDK, not Tendermint), but it's something to consider.

As part of all of this decision making, though, it'd be nice to see some thorough benchmark results with each DB Backend type in question. Ideally, it'd be broken down into several benchmarks, such as one just for reading, one just for writing, a few with different mixes of reading and writing (e.g. 25% read 75% write, 50/50, and 75/25). Disk footprint should also be recorded with those benchmarks. I'd also like to see the benchmarks starting with a fresh db vs starting with a DB with 100G of data in it already.

I feel like that information would allow for better decisions than anything.

**Comment by creachadair** [comment]
@dwedul-figure I broadly agree with most of these points. There are a few specific things I think are worth highlighting in this design space:

1. Both storage footprint and write latency are more directly affected by our choice of which data we store, and how we organize them into keys and values, than by the I/O performance of the database backend (but see point (2) also).

   Right now, regardless of which database backend is configured, we store data in a fairly naive way (typically raw serialized protobuf) and very little effort is made to deduplicate or remove unneeded fields before storage. To some extent this is forced upon us by the generic API we're using to store things, but mostly we need to revisit how we represent our data in storage vs. other parts of the protocol. Those issues are very much in-scope for our upcoming database plans.

   I believe we should be able to substantially reduce both storage footprint and I/O load by eliminating these baked-in redundancies and better-partitioning data so that we don't have to read entire large blobs just to access individual fields. This is pretty well-trodden ground in other projects, and Tendermint has a lot of low-hanging fruit to pick in this context.

2. Our various storage tables (e.g., blockstore vs. index vs. statestore) have very different load profiles. Index queries in older nodes (v0.34 and earlier) are very likely to be dominated by the inability to perform efficient range queries (thus, queries wind up scanning the entire store for a given type). The migration to use ordered codes for keys in v0.35.x should have already helped substantially with that problem, even without a database change. (And the addition of the PSQL indexing sink provides another escape hatch for query-intensive workloads)

   That said, indexing is also much more write-intensive than block storage, and we've seen some traces that suggest there _are_ real write throughput issues for this path. Much of this is down to suboptimal design of the KV indexer. We have bigger plans for indexing and query APIs that we're considering (see #7135 for some good starting points), and I believe we can best address most of the reported performance issues with indexing by starting at a higher level.

To sum up, reiterating what @tychoish has mentioned, our choice of which database to use does have performance implications, but it isn't the long pole at this point. For the workloads internal to the consensus node, I expect we will get much bigger dividends by tackling more fundamental storage design issues in Tendermint itself.

Of course, none of this addresses other issues applications may have, and I could imagine apps might still benefit from being able to swap out their database backend for idiosyncratic workloads. That's mostly orthogonal to this issue, however: This is really _only_ about what the consensus node does internally.

**Comment by faddat** [comment]
https://github.com/tendermint/tm-db/issues/260

Just an update from my end-- it looks like badger won't be actively developed. 

So if we break it down to what is in native Go, and what will actively be developed, I think we are more or less left with:

* pebble - super active
* bolt - not super active (but better than goleveldb)
* goleveldb - not super active

I've heard a bunch of what I believe to be very legitimate concern about performance using native rocks that relate to cgo ffi. 

I am also going to try and shimmy pebble support into the cosmos-sdk this month.  No guarantees, since I had stability issues there when trying to use pebble in prodcution on both relay and public facing rpc nodes. 

Cheers!

**Comment by dwedul-figure** [comment]
Boltdb does not scale.

A couple months ago, I did some testing using our Cosmos-SDK v0.45 based Provenance Blockchain. Using `boltdb` as the backend for all databases resulted in block replay times double that of block creation times. There is no way a node using `boltdb` as the db backend (for all cosmos and tm dbs) would ever catch up.

In some timing tests I did (using some sim tests), I actually found that `goleveldb` was the fastest, then `badgerdb`, then `cleveldb`, then `rocksdb`. I didn't even try to time `boltdb`.

I do have an extra concern with `rocksdb` too. We have some github actions that run the sim tests using various db backends. The ones that use `rocksdb` sporadically fail at the start due to an 'Illegal Instruction' error. I've tried many versions of `rocksdb` and they all run into that.
